src/semidiscretization/semidiscretization_coupled_p4est.jl

```diff
@@ -0,0 +1,545 @@
+        summary_header(io, "System #$i")
+
+        summary_line(io, "mesh", mesh |> typeof |> nameof)
+        show(increment_indent(io), MIME"text/plain"(), mesh)
+
+        summary_line(io, "equations", equations |> typeof |> nameof)
+        show(increment_indent(io), MIME"text/plain"(), equations)
+
+        summary_line(io, "solver", solver |> typeof |> nameof)
+        show(increment_indent(io), MIME"text/plain"(), solver)
+
+        summary_footer(io)
+        println(io, "\n")
+    end
+end
+
+@inline Base.ndims(semi::SemidiscretizationCoupledP4est) = ndims(semi.semis[1])
+
+@inline nsystems(semi::SemidiscretizationCoupledP4est) = length(semi.semis)
+
+@inline eachsystem(semi::SemidiscretizationCoupledP4est) = Base.OneTo(nsystems(semi))
+
+@inline Base.real(semi::SemidiscretizationCoupledP4est) = promote_type(real.(semi.semis)...)
+
+@inline function Base.eltype(semi::SemidiscretizationCoupledP4est)
+    promote_type(eltype.(semi.semis)...)
+end
+
+@inline function ndofs(semi::SemidiscretizationCoupledP4est)
+    sum(ndofs, semi.semis)
+end
+
+"""
+    ndofsglobal(semi::SemidiscretizationCoupledP4est)
+
+Return the global number of degrees of freedom associated with each scalar variable across all MPI ranks, and summed up over all coupled systems.
+This is the same as [`ndofs`](@ref) for simulations running in serial or
+parallelized via threads. It will in general be different for simulations
+running in parallel with MPI.
+"""
+@inline function ndofsglobal(semi::SemidiscretizationCoupledP4est)
+    sum(ndofsglobal, semi.semis)
+end
+
+function compute_coefficients(t, semi::SemidiscretizationCoupledP4est)
+    @unpack u_indices = semi
+
+    u_ode = Vector{real(semi)}(undef, u_indices[end][end])
+
+    for i in eachsystem(semi)
+        # Call `compute_coefficients` in `src/semidiscretization/semidiscretization.jl`
+        u_ode[u_indices[i]] .= compute_coefficients(t, semi.semis[i])
+    end
+
+    return u_ode
+end
+
+@inline function get_system_u_ode(u_ode, index, semi::SemidiscretizationCoupledP4est)
+    @view u_ode[semi.u_indices[index]]
+end
+
+function rhs!(du_ode, u_ode, semi::SemidiscretizationCoupledP4est, t)
+    time_start = time_ns()
+
+    n_nodes = length(semi.semis[1].mesh.parent.nodes)
+    # Reformat the global solutions vector.
+    u_ode_reformatted = Vector{real(semi)}(undef, ndofs(semi))
+    u_ode_reformatted_reshape = reshape(u_ode_reformatted,
+                                        (n_nodes,
+                                         n_nodes,
+                                         length(semi.mesh_ids)))
+    # Extract the global solution vector from the local solutions.
+    foreach_enumerate(semi.semis) do (i, semi_)
+        system_ode = get_system_u_ode(u_ode, i, semi)
+        system_ode_reshape = reshape(system_ode,
+                                     (n_nodes, n_nodes,
+                                      Int(length(system_ode) / n_nodes^2)))
```

Comment:
```suggestion
                                      Int(length(system_ode) / n_nodes^2)))
```

The `^2` looks 2d hard-coded to me. What about `ndims(mesh)`?

Reply:
Hm, one less thing to change when updating to 3d, right?